In [771]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt         
%matplotlib inline

In [772]:
# Load data from pickle file (9010, 31306) last 30 Years
data_sp = pd.read_pickle("C:/Users/Carap/OneDrive/Escritorio/Proyecto/datasets/data componentes SP500.pkl")

In [773]:
data_sp_copy = data_sp

In [774]:
data_sp_copy.shape

(9010, 31306)

In [775]:
# Definir Año para filtrar
start_date  =  '2015-01-01'
data_sp['date'] = data_sp_copy.index
data_sp_copy.reset_index(drop=True)

0111145D UN Equity                                                     \
                   open    high     low   close    volume bid ask open_int   
0                44.750  45.750  44.750  45.000  264800.0 NaN NaN      NaN   
1                45.250  45.500  45.000  45.125   12100.0 NaN NaN      NaN   
2                45.375  46.000  45.125  46.000   21100.0 NaN NaN      NaN   
3                   NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
4                   NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
5                   NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
6                46.500  47.000  46.125  46.875   59400.0 NaN NaN      NaN   
7                46.375  46.500  46.125  46.500   89200.0 NaN NaN      NaN   
8                46.000  46.500  45.000  45.250  114200.0 NaN NaN      NaN   
9                45.000  45.125  44.625  44.750   69800.0 NaN NaN      NaN   
10               44.625  44.750  43.875  44.000  118100.0 NaN NaN      NaN   
11               44.250  44.250  43.750  43.750   58600.0 NaN NaN      NaN   
12               43.625  43.875  43.625  43.750   85900.0 NaN NaN      NaN   
13               44.000  44.500  43.375  43.375   64000.0 NaN NaN      NaN   
14               42.875  43.000  42.000  42.250  112800.0 NaN NaN      NaN   
15               42.125  42.125  41.875  41.875   16700.0 NaN NaN      NaN   
16               41.875  42.000  41.500  41.875   40600.0 NaN NaN      NaN   
17               42.000  42.125  41.500  41.750   19900.0 NaN NaN      NaN   
18               41.500  41.875  41.250  41.625   39800.0 NaN NaN      NaN   
19               41.875  42.000  41.250  41.250   33500.0 NaN NaN      NaN   
20               41.250  41.375  39.500  39.500   52700.0 NaN NaN      NaN   
21               39.625  40.125  39.500  40.000   34300.0 NaN NaN      NaN   
22               39.500  40.125  39.125  39.625   30300.0 NaN NaN      NaN   
23               39.875  40.250  39.625  39.625   45600.0 NaN NaN      NaN   
24               39.750  40.375  39.750  40.375   42000.0 NaN NaN      NaN   
25                  NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
26                  NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
27               40.625  40.625  40.125  40.375   22800.0 NaN NaN      NaN   
28               40.625  40.625  40.125  40.250   35100.0 NaN NaN      NaN   
29               40.625  41.375  40.625  41.375   24600.0 NaN NaN      NaN   
...                 ...     ...     ...     ...       ...  ..  ..      ...   
8980                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8981                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8982                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8983                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8984                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8985                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8986                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8987                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8988                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8989                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8990                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8991                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8992                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8993                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8994                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8995                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8996                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8997                NaN     NaN     NaN     NaN       NaN NaN NaN      NaN   
8998                NaN    

In [776]:
# Aqui se filtra
data_sp_copy['date']=pd.to_datetime(data_sp_copy['date'])
datefilter  =  (data_sp_copy['date']  >  start_date )
data_sp_copy  =  data_sp_copy.loc[datefilter] 

In [777]:
# resetear index
data_sp_copy.index = range(data_sp_copy.shape[0])

In [779]:
# guardar date en dataframe 
date = pd.DataFrame(data_sp_copy['date'])

In [780]:
# extraer nombres de compañias
company=[]
for i in data_sp_copy.columns:
    company.append(i[0])

In [782]:
company = list(dict.fromkeys(company))

In [783]:
def check(word, list):
    if word in list:
        print("word is in the list!")
    else:
        print("word is not in the list!")
check('date', company)

word is in the list!


In [784]:
# elimino date
removeWord = ['date']
company = [word for word in company if word not in removeWord]
check('date', company)

word is not in the list!
word is not in the list!


In [787]:
# creo una lista con cada compania como dataframe
dataframes = []
for i in company:
    dataframes.append(data_sp_copy[i])


In [789]:
# creo la columna company con  cada nombre de la compañia
j=0
for i in company:
    dataframes[j]['Company'] = i
    dataframes[j]['Date'] = date['date']
    j += 1

C:\Users\Carap\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
C:\Users\Carap\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [792]:
# Crear una variable para cada dataframe
# for i, val in enumerate(dataframes): 
#     globals()["dataframe_Sep_%d"%i] = val

# **TODO BIEEEEEENNNN**

In [795]:
test = dataframes.copy()

# **ENSEÑARRRRRRR**

In [819]:
## Revisar por que se tiene que ranear 5 veces aprox.
j = 0
for i in test:
    if i['open'].isnull().mean() < 0.20:
        pass        
    else:
        del test[j]
    j += 1

In [830]:
j = 0
for i in test:
    test[j]['Profitability'] = (test[j]['close']-test[j]['open'])/test[j]['open']*100
    j += 1

C:\Users\Carap\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [835]:
test[0].tail(500)

open    high    low  close     volume    bid    ask  cur_mkt_cap  \
846     NaN     NaN    NaN    NaN        NaN    NaN    NaN          NaN   
847   69.22  69.490  69.17  69.31   415141.0  69.32  69.33   22317.8200   
848   69.56  70.920  69.56  70.59   649255.0  70.57  70.58   22729.9800   
849   70.50  70.840  68.45  68.69   552656.0  68.68  68.69   22118.1800   
850   68.97  69.590  68.25  69.24   769535.0  69.22  69.23   22295.2800   
851   69.19  69.990  67.23  68.69   653356.0  68.67  68.68   22118.1800   
852   68.94  69.170  66.19  66.24   587653.0  66.23  66.24   21329.2800   
853   66.20  67.000  65.73  65.83   790607.0  65.80  65.81   21197.2600   
854   65.88  66.250  65.62  65.89   531068.0  65.89  65.90   21216.5800   
855   65.91  66.860  65.77  66.71   890913.0  66.71  66.72   21480.6200   
856   66.97  67.520  66.68  67.38   764180.0  67.37  67.38   21696.3600   
857   67.43  67.710  66.48  67.00   683723.0  66.95  66.96   21574.0000   
858   67.00  67.155  66.52  67.06   514013.0  67.04  67.05   21593.3200   
859   67.23  67.650  66.64  66.68   378336.0  66.68  66.69   21470.9600   
860   66.68  67.040  66.40  66.51   567792.0  66.52  66.53   21416.2200   
861   66.97  67.850  66.69  67.61  1545185.0  67.60  67.61   21770.4200   
862   67.89  68.500  67.48  67.66   941786.0  67.62  67.63   21786.5200   
863   67.94  68.020  67.51  67.75   623910.0  67.65  67.66   21815.5000   
864   67.81  68.130  67.16  67.40   441364.0  67.35  67.36   21702.8000   
865   67.51  67.870  67.28  67.52   557680.0  67.49  67.50   21810.1772   
866   67.41  67.530  66.86  67.35   324950.0  67.31  67.32   21755.2641   
867   67.31  67.420  67.05  67.25   198765.0  67.23  67.24   21722.9623   
868   67.21  67.460  67.15  67.30   262594.0  67.30  67.31   21739.1132   
869   67.44  67.480  67.05  67.45   157457.0  67.45  67.46   21787.5659   
870   67.50  67.550  66.93  66.97   518923.0  66.94  66.95   21632.5173   
871     NaN     NaN    NaN    NaN        NaN    NaN    NaN          NaN   
872     NaN     NaN    NaN    NaN        NaN    NaN    NaN          NaN   
873   67.45  67.870  67.34  67.60   314624.0  67.60  67.61   21836.0186   
874   67.62  69.490  67.62  69.32   416378.0  69.35  69.36   22391.6096   
875   69.54  69.810  68.80  68.80   586767.0  68.80  68.81   22223.6403   
...     ...     ...    ...    ...        ...    ...    ...          ...   
1316  70.05  70.320  69.13  69.48   499381.0  69.48  69.50   21538.8000   
1317  69.26  70.250  68.81  69.94   323281.0  69.94  69.96   21681.4000   
1318  70.86  71.285  70.48  70.70   366453.0  70.70  70.72   21917.0000   
1319  71.45  72.340  70.82  71.11   861947.0  71.06  71.07   22006.2466   
1320  70.70  71.040  69.51  70.41   509611.0  70.41  70.43   21789.6192   
1321  70.94  71.220  69.96  70.86   610082.0  70.84  70.86   21928.8797   
1322  71.64  73.700  71.64  73.52   877392.0  73.53  73.54   22752.0637   
1323  73.69  74.370  73.69  74.28   603119.0  74.28  74.29   22987.2591   
1324  74.44  74.590  73.69  74.18   467152.0  74.16  74.18   22956.3124   
1325  73.50  75.020  73.30  74.77   506229.0  74.77  74.79   23138.8983   
1326  75.17  77.000  75.17  76.99   576378.0  76.98  77.01   23825.9165   
1327  77.01  78.090  76.71  77.57   546638.0  77.57  77.58   24005.4078   
1328  77.56  78.050  77.27  77.90   365200.0  77.88  77.89   24107.5321   
1329  77.32  77.420  76.51  77.15   436987.0  77.10  77.12   23875.4314   
1330  76.95  76.950  76.33  76.47   133992.0  76.44  76.48   23757.8326   
1331  76.36  77.570  75.75  77.57   518311.0  77.52  77.56   24005.4078   
1332  77.58  77.660  77.41  77.53    17477.0  77.51  77.56          NaN   
1333  78.22  78.720  77.92  77.97  2143728.0  77.92  77.93   24129.1949   
1334  77.48  78.370  77.32  78.06   312279.0  78.05  78.06   24157.0469   
1335  78.29  78.390  75.64  75.97   700584.0  75.94  75.97   23510.2595   
1336  76.05  76.540  75.16  76.33   314275.0  76.33  76.34   23621.6679   
1337  76.45  76.940  76.11  76.5

In [ ]:
close - open / open

In [660]:
dataframe1.columns

Index(['open', 'high', 'low', 'close', 'volume', 'bid', 'ask', 'cur_mkt_cap',
       'pe_ratio', 'tot_debt_to_tot_asset', 'cf_free_cash_flow',
       'return_tot_eqy', 'eqy_sh_out', 'eqy_dvd_yld_ind', 'ebitda',
       'return_on_inv_capital', 'return_on_asset', 'return_on_cap',
       'return_com_eqy', 'net_income', 'short_and_long_term_debt',
       'sales_rev_turn', 'gross_profit', 'announcement_dt', 'total_equity',
       'close_adj', 'Company'],
      dtype='object')

In [864]:
import datetime
date['date'] = pd.to_datetime(date['date'])
#date_list = list(date['date'])
#date_list.weekday()

AttributeError: 'list' object has no attribute 'weekday'

In [847]:
datelist.get_weekday()

AttributeError: 'list' object has no attribute 'get_weekday'

# **Siguientes Pasos:**
## Mirar si estan todos los dias o cuales faltan
## Crear columna weekday
    * sabados y domingos  rentabilidad 0
    * eliminar los Nas de rentabilidad superior de 15% de NAs (funcion de revisar)
    * aplicar medias moviles exponenciales
## Crear columna Mes y Año
## Rentabilidad media por Mes y la desviacion standard por mes